File: crates/nmp-wasm/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WorkerRequest {
    Hello(ClientHello),
    Start(StartConfig),
    /// ADR-0063 reference-resolution control message.
    ///
    /// This is deliberately separate from the app-write doorway. It carries the
    /// raw-key `resolve_ref` fields as typed JSON data so components can mount
    /// profile/event interests without reopening the retired `action_type +
    /// Value` command surface.
    ResolveRef(ResolveRef),
    /// ADR-0063 reference-release control message.
    ReleaseRef(ReleaseRef),
    /// ADR-0064 / S2 (#1750) — the **binary write doorway**. The host posts the
    /// raw bytes of a finished `DispatchEnvelope` (`correlation_id` + generated
    /// `action_namespace` + typed FlatBuffers `payload`). This is the SAME open
    /// transport the native FFI (`nmp_app_dispatch_action_bytes`) uses — there
    /// is no wasm-specific write vocabulary (the hand-rolled `AppAction` enum +
    /// `"app_action"` envelope were deleted in #1743 Cut A).
    DispatchBytes(DispatchBytes),
    CapabilityResult(CapabilityResult),
    /// Set the active identity for app-level writes.
    ///
    /// The browser host runs the asynchronous half of the handshake itself
    /// (e.g. `await window.nostr.getPublicKey()` for NIP-07) and supplies the
    /// already-known pubkey hex in this request. The wasm runtime validates +
    /// canonicalizes the pubkey and seeds the kernel's active account so
    /// active-follows resolution and bootstrap interests know whose data to
    /// load.
    ///
    /// **No persistent signer is installed** (ADR-0064 §5): signing is the
    /// ADR-0050 capability round-trip ([`Self::BeginSign`] →
    /// [`WorkerEvent::SignRequest`] → [`Self::DeliverSignerResponse`]), not an
    /// `Arc<dyn Signer>` awaited inside the publish flow. `kind`: `"nip07"` is
    /// the only kind wired; other kinds return [`WorkerEvent::CapabilityFailure`]
    /// with `unsupported_signer_kind`.
    SetIdentity(SetIdentity),
    /// #1753 S6 — begin a NIP-07 sign capability round-trip.
    ///
    /// The worker parks a sign op (ADR-0050 §D1) bound to `account_pubkey` and
    /// emits a [`WorkerEvent::SignRequest`] the **main-thread JS bridge** (the
    /// broker — Web Workers have no `window.nostr`) fulfils by calling
    /// `window.nostr.signEvent(unsigned)`. The bridge posts the signed bytes
    /// back as [`WorkerRequest::DeliverSignerResponse`]. Pure message re-entry:
    /// no polling, no tick-dependence (D8).
    BeginSign(BeginSign),
    /// #1753 S6 — the main-thread broker delivers a signer response (the
    /// `sign`-verb fulfiller feeding the ADR-0050 §D3b `DeliverSignerResponse`
    /// command). `signed_json` is the flat-NIP-01 event from
    /// `window.nostr.signEvent`; `error` is set instead when the user rejected
    /// or `window.nostr` was unavailable. Account-pinned: the worker rejects a
    /// signature authored by an account other than the one the round-trip was
    /// begun for.
    DeliverSignerResponse(DeliverSignerResponse),
    Stop {
        correlation_id: String,
    },
}

/// Payload for [`WorkerRequest::BeginSign`].
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct BeginSign {
    /// Lowercase-hex account the sign is pinned to (the host already has it from
    /// the NIP-07 `getPublicKey()` handshake).
    pub account_pubkey: String,
    /// The unsigned flat-NIP-01 (or nested-`UnsignedEvent`) JSON to sign.
    pub unsigned_json: String,
}

/// Payload for [`WorkerRequest::DeliverSignerResponse`].
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct DeliverSignerResponse {
    /// The `correlation_id` from the [`WorkerEvent::SignRequest`] this fulfils.
    pub correlation_id: String,
    /// The signed flat-NIP-01 event JSON from `window.nostr.signEvent`. Present
    /// on success; `None` when `error` is set.
    #[serde(default)]
    pub signed_json: Option<String>,
    /// A failure reason (user rejected, no `window.nostr`, etc.). Present on
    /// failure; `None` on success.
    #[serde(default)]
    pub error: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ClientHello {
    pub app_id: String,
    pub platform: String,
    pub protocol_version: u16,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct StartConfig {
    pub app_id: String,
    /// Relay set the host wants the runtime to connect to. Relay policy is a
    /// host concern, not framework policy — the framework has no built-in
    /// default. A host (e.g. the Chirp web composition root) MUST supply this
    /// explicitly; a missing field fails serde deserialization loudly rather
    /// than silently falling back to one app's relays.
    pub relays: Vec<String>,
    /// Explicit relay bootstrap list (url + role). Like `relays`, this is host
    /// policy with no framework default; the host MUST supply it.
    pub relay_bootstrap: Vec<RelayBootstrapEntry>,
    pub database_name: String,
    pub correlation_id: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct RelayBootstrapEntry {
    pub url: String,
    pub role: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ResolveRef {
    pub namespace: u32,
    pub key: String,
    pub consumer_id: String,
    pub shape: u32,
    pub liveness: u32,
    /// Optional relay hints decoded by the app boundary from a NIP-19/NIP-21
    /// event reference. Missing defaults to the bare-key path.
    #[serde(default)]
    pub hints: Vec<String>,
    /// Optional event author decoded by the app boundary from a nevent author
    /// TLV. Ignored for profile refs and address-coordinate event keys, where
    /// the author is already part of the raw key.
    #[serde(default)]
    pub event_author: Option<String>,
    pub correlation_id: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReleaseRef {
    pub namespace: u32,
    pub key: String,
    pub consumer_id: String,
    pub correlation_id: String,
}

/// Payload for [`WorkerRequest::DispatchBytes`] — the ADR-0064 binary write
/// doorway. `bytes` are the raw bytes of a finished `DispatchEnvelope`
/// FlatBuffers root (file identifier `NMPD`); the runtime decodes them through
/// `nmp_core::dispatch_envelope::decode_dispatch_envelope` (the SAME path the
/// native FFI `nmp_app_dispatch_action_bytes` uses). The host posts them as a
/// transferable `Uint8Array`; serde renders them as a JSON number array on the
/// degraded in-process path.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct DispatchBytes {
    pub bytes: Vec<u8>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct CapabilityResult {
    pub capability: String,
    pub correlation_id: String,
    pub payload: Value,
}

/// Payload for [`WorkerRequest::SetIdentity`].
///
/// `kind` is the backend discriminator the host obtained the pubkey from.
/// `"nip07"` is the only kind wired; other kinds are honestly rejected rather
/// than silently dropped. The runtime seeds the kernel's active account from
/// the validated pubkey — it does NOT install a persistent signer (ADR-0064 §5:
/// signing is the [`WorkerRequest::BeginSign`] capability round-trip).
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SetIdentity {
    /// Backend kind. Currently must be `"nip07"`.
    pub kind: String,
    /// Hex-encoded public key the host already obtained from the backend.
    ///
    /// For NIP-07 this is the result of `await window.nostr.getPublicKey()`.
    /// Supplied by the host so the wasm runtime stays synchronous — the async
    /// getPublicKey() round-trip happens in JS, before the request is sent.
    pub pubkey_hex: String,
    /// Correlation id echoed back in [`WorkerEvent::ActionAccepted`] (or
    /// [`WorkerEvent::CapabilityFailure`] on failure) so the host can match
    /// the outcome to the request that triggered it.
    pub correlation_id: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct CapabilityFailure {
    pub capability: String,
    pub correlation_id: String,
    pub reason: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RuntimeStatus {
    Ready,
    Running,
    Degraded(DegradedMode),
    Stopped,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DegradedMode {
    BrowserActorDriverMissing,
    CapabilityRejected,
    ProtocolMismatch,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WorkerEvent {
    HelloAccepted {
        protocol_version: u16,
        status: RuntimeStatus,
    },
    RuntimeStatus {
        status: RuntimeStatus,
        correlation_id: Option<String>,
    },
    ActionAccepted {
        action_type: String,
        correlation_id: String,
    },
    UpdateBytes {
        bytes: Vec<u8>,
    },
    CapabilityFailure(CapabilityFailure),
    /// #1753 S6 — a sign capability request the worker emits for the
    /// **main-thread JS bridge** (the broker) to fulfil. The bridge calls
    /// `window.nostr.signEvent(unsigned_json)` (ensuring `window.nostr` is on
    /// `account_pubkey` first) and posts the result back as a
    /// [`WorkerRequest::DeliverSignerResponse`] carrying this `correlation_id`.
    SignRequest {
        correlation_id: String,
        account_pubkey: String,
        unsigned_json: String,
    },
    /// #1753 S6 — a sign round-trip completed via message re-entry. Carries the
    /// signed flat-NIP-01 JSON. For publish-dispatch round-trips the wasm
    /// runtime also consumes the same completion to call `publish_pre_signed`
    /// and fan outbound frames; the JS host never publishes inline.
    SignCompleted {
        correlation_id: String,
        signed_json: String,
    },
    /// #1753 S6 — a sign round-trip failed (parse error, account-pin mismatch,
    /// user rejection, or an unknown/stale correlation id).
    SignFailed {
        correlation_id: String,
        reason: String,
    },
    Error {
        code: String,
        message: String,
        correlation_id: Option<String>,
    },
}

/// Normalise the relay bootstrap list from a [`StartConfig`]: if
/// `relay_bootstrap` is non-empty, it is the authoritative list; otherwise
/// synthesise one `RelayBootstrapEntry` per URL string in `relays` with
/// `role = "both"`.
pub(crate) fn relay_bootstrap_from_config(
    relays: Vec<String>,
    relay_bootstrap: Vec<RelayBootstrapEntry>,
) -> Vec<RelayBootstrapEntry> {
    if !relay_bootstrap.is_empty() {
        return relay_bootstrap;
    }
    relays
        .into_iter()
        .map(|url| RelayBootstrapEntry {
            url,
            role: "both".to_string(),
        })
        .collect()
}

File: web/packages/runtime-web/src/protocol.ts
export type WorkerRequest =
  | { type: "hello"; app_id: string; platform: string; protocol_version: number }
  | {
      type: "start";
      app_id: string;
      /** Relay set the host wants the runtime to connect to. Relay policy is a
       *  host concern (#1125): the framework defines no defaults, so the host
       *  MUST supply this. */
      relays: string[];
      /** Explicit relay bootstrap list (url + role). Host policy with no
       *  framework default — the host MUST supply it. When non-empty the wasm
       *  runtime uses it verbatim and ignores `relays`. */
      relay_bootstrap: { url: string; role: string }[];
      database_name: string;
      correlation_id: string;
    }
  /** ADR-0063 structured reference-resolution control. This is not an app
   *  write doorway and cannot carry arbitrary action namespaces. */
  | {
      type: "resolve_ref";
      namespace: number;
      key: string;
      consumer_id: string;
      shape: number;
      liveness: number;
      /** Optional relay hints decoded from NIP-19/NIP-21 event refs. */
      hints?: string[];
      /** Optional nevent author TLV decoded at the app boundary. */
      event_author?: string | null;
      correlation_id: string;
    }
  /** ADR-0063 structured reference release. */
  | {
      type: "release_ref";
      namespace: number;
      key: string;
      consumer_id: string;
      correlation_id: string;
    }
  /** ADR-0064 / S2 (#1750) — the typed binary write doorway. `bytes` are a
   *  finished `DispatchEnvelope` FlatBuffers root (file id `NMPD`) carrying the
   *  correlation_id + generated action_namespace + opaque typed payload. This
   *  is the ONLY app-level write path (the hand-rolled `app_action` envelope was
   *  deleted in #1743 Cut A); it is identical in shape to the native FFI seam. */
  | {
      type: "dispatch_bytes";
      bytes: Uint8Array;
    }
  /** NIP-50 public search. The worker validates/bounds the query through
   *  `nmp_nip50::SearchRequest` and emits typed results under
   *  `nmp.nip50.search.<session_id>` (`N50S`) in update snapshots. */
  | {
      type: "search_open";
      session_id: string;
      query: string;
      scope: "notes" | "profiles" | "longform";
      targets: "user_preferred" | "app_default" | "explicit";
      relays?: string[];
      max_hits?: number;
      correlation_id: string;
    }
  | {
      type: "search_close";
      session_id: string;
      correlation_id: string;
    }
  /** NIP-29 public group discovery. The worker owns the relay-pinned metadata
   *  interest and emits `nmp.nip29.discovered_groups` (`NDGS`) in snapshots. */
  | {
      type: "group_discovery_open";
      session_id: string;
      relay_url: string;
      correlation_id: string;
    }
  | {
      type: "group_discovery_close";
      session_id: string;
      correlation_id: string;
    }
  /** Browser runtime relay inventory edit. This is structured transport/runtime
   *  control, not an app-level write. The Rust runtime validates URL/role,
   *  mutates the configured-relay projection, and opens/closes browser relay
   *  drivers so diagnostics reflect the live socket inventory. */
  | {
      type: "relay_config";
      action: "add" | "remove";
      url: string;
      role?: string;
      correlation_id: string;
    }
  /** Publish the Rust-owned configured relay projection as NIP-65 kind:10002.
   *  This is a typed write: the worker encodes `nmp.nip65.publish_relay_list`
   *  and routes it through the same `dispatch_bytes` path, so TS never builds
   *  NIP-65 tags or marker policy. */
  | {
      type: "publish_relay_preferences";
      correlation_id: string;
    }
  | {
      type: "capability_result";
      capability: string;
      correlation_id: string;
      payload: unknown;
    }
  | { type: "stop"; correlation_id: string }
  /** Set the active identity. Signer kinds:
   *
   *  kind: "nip07" — the browser host called window.nostr.getPublicKey() first
   *    and provides pubkey_hex. Signing is a begin_sign capability round-trip
   *    routed back to the main thread (ADR-0064 §5). Does NOT install a
   *    persistent signer. This is the path the current nmp-browser-runtime
   *    wires end-to-end.
   *
   *  kind: "local_key" — the host provides secret_key_bech32 (the raw nsec
   *    bech32 string) and an empty pubkey_hex; the runtime is responsible for
   *    decoding the nsec, deriving the pubkey, and installing a LocalKey
   *    provider. The TS layer MUST NOT decode or sign with the nsec — handing
   *    it to the runtime is the only permitted use.
   */
  | {
      type: "set_identity";
      kind: string;
      /** Hex pubkey. Carries the NIP-07 pubkey for kind: "nip07"; sent as an
       *  empty string for kind: "local_key" (the runtime derives it from the
       *  secret key). Always present so the request deserializes against the
       *  runtime's required field. */
      pubkey_hex: string;
      /** For kind: "local_key" — raw nsec bech32 string handed to the runtime's
       *  LocalKey provider. Never decoded or signed-with in TS. */
      secret_key_bech32?: string;
      correlation_id: string;
      /** Raw relay permissions reported by the identity backend, e.g. NIP-07
       *  getRelays(). The worker owns canonicalization and role mapping. */
      identity_relays?: IdentityRelayPermission[];
    }
  /** #1753 S6 — begin a NIP-07 sign capability round-trip. The worker parks a
   *  sign op (ADR-0050 §D1) and emits a `sign_request` event the MAIN THREAD
   *  fulfils via window.nostr.signEvent (Workers have no window.nostr). Pure
   *  message re-entry: no polling (D8). */
  | {
      type: "begin_sign";
      account_pubkey: string;
      unsigned_json: string;
    }
  /** #1753 S6 — the main-thread broker delivers the signer response (the
   *  `sign`-verb fulfiller feeding ADR-0050 §D3b). Exactly one of `signed_json`
   *  / `error` is set. Account-pinned: the worker rejects a signature authored
   *  by a different account than the round-trip was begun for. */
  | {
      type: "deliver_signer_response";
      correlation_id: string;
      signed_json?: string | null;
      error?: string | null;
    }
  /** Read-only diagnostics request: return the kernel-owned bounded routing
   *  trace JSON. This is not a write doorway and carries no routing policy in
   *  the host; the worker calls the wasm runtime's existing
   *  `recent_routing_decisions()` accessor. */
  | {
      type: "routing_decisions";
      correlation_id: string;
    };

export type RuntimeStatus =
  | "ready"
  | "running"
  | "stopped"
  | {
      degraded:
        | "browser_actor_driver_missing"
        | "browser_bridge_unavailable"
        | "capability_rejected"
        | "protocol_mismatch";
    };

export type IdentityRelayPermission = {
  url: string;
  read?: boolean;
  write?: boolean;
};

export type WorkerEvent =
  | { type: "hello_accepted"; protocol_version: number; status: RuntimeStatus }
  | { type: "runtime_status"; status: RuntimeStatus; correlation_id?: string }
  | { type: "action_accepted"; action_type: string; correlation_id: string }
  | { type: "update_bytes"; bytes: Uint8Array }
  | {
      type: "capability_failure";
      capability: string;
      correlation_id: string;
      reason: string;
    }
  /** #1753 S6 — a sign capability request the worker emits for the MAIN-THREAD
   *  broker to fulfil: call window.nostr.signEvent(unsigned_json) (ensuring the
   *  extension is on account_pubkey first) and post back a
   *  `deliver_signer_response` carrying this correlation_id. */
  | {
      type: "sign_request";
      correlation_id: string;
      /** Present when this sign request was caused by an action dispatch; lets
       *  the UI settle the initiating command while the signer broker continues. */
      action_correlation_id?: string | null;
      account_pubkey: string;
      unsigned_json: string;
    }
  /** #1753 S6 — a sign round-trip completed via message re-entry. For publish
   *  dispatches the worker also consumes this completion to publish through the
   *  Rust kernel; the main thread never publishes inline. */
  | { type: "sign_completed"; correlation_id: string; signed_json: string }
  /** #1753 S6 — a sign round-trip failed (parse, account-pin mismatch, user
   *  rejection, or unknown/stale correlation id). */
  | { type: "sign_failed"; correlation_id: string; reason: string }
  /** #968 — read-only routing diagnostics snapshot, JSON rendered by Rust from
   *  the kernel-owned `RoutingTraceProjection`. */
  | { type: "routing_decisions"; correlation_id: string; json: string }
  | { type: "error"; code: string; message: string; correlation_id?: string };

export const protocolVersion = 1;

export function eventCorrelationId(event: WorkerEvent): string | undefined {
  switch (event.type) {
    case "runtime_status":
    case "action_accepted":
    case "error":
      return event.correlation_id;
    case "capability_failure":
      return event.correlation_id;
    // #1753 S6 — sign round-trip terminals are correlation-keyed so a pending
    // request resolves on completion; `sign_request` is NOT (it is a broker
    // instruction the main thread acts on, not a reply to a pending caller).
    case "sign_completed":
    case "sign_failed":
    case "routing_decisions":
      return event.correlation_id;
    case "sign_request":
    case "hello_accepted":
    case "update_bytes":
      return undefined;
  }
}

export function labelRuntimeStatus(status: RuntimeStatus): string {
  if (typeof status === "string") {
    return status.replace("_", " ");
  }
  return status.degraded.split("_").join(" ");
}
